### PR TITLE
Feature/star stringtie rnaseq

### PR DESCRIPTION
**Pull request by ens-carlos**
# Requirements
When creating your Pull request, please fill out the template below:

# PR details
_Is this a fix/ update/ new feature?_
New feature
_Include a short description_
New subpipeline config for star stringtie rnaseq
_Include links to JIRA tickets_
https://www.ebi.ac.uk/panda/jira/browse/ENSGENEBUI-485
https://www.ebi.ac.uk/panda/jira/browse/ENSGENEBUI-490
# Testing
_Have you tested it?_
perl -c
# Assign to the weekly GitHub reviewer
_If you are a member of Ensembl, please check the Genebuild weekly Rotas and assign this week's GitHub reviewer to the PR_

### CODE REVIEW

**Review by thibauthourlier — CHANGES_REQUESTED** [review]
Two files should not be in the PR, Genome_anntotation_conf.pm and UniProtCladeDownloadStatic. The changes need to be reverted to their status in feature/breaking_down_1
The copyright value for most of the file is wrong to me, there should be two copyright lines. HiveStar2Introns has the correct values.
It would be great if you can add POD to the methods and check that they are at least correct where the POD is present. I can help you with that if needed.
I feel really bad because I didn't give you the correct information for the pipeline config. I thought the starting branch will only have stringtie and not a mix of everything. I should have checked the non vert pipeline. I will message you to see what is the best way forward.

**Comment by thibauthourlier** [comment]
I'll probably just use the config later if it's needed for the non vert pipeline. Otherwise it's better to use PR208 to generate the stringtie config

File: modules/Bio/EnsEMBL/Analysis/Hive/RunnableDB/GenerateGTFLoadingJobs.pm [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+use parent ('Bio::EnsEMBL::Analysis::Hive::RunnableDB::HiveBaseRunnableDB');
+
+
+sub param_defaults {
+  my ($self) = @_;
+
+  return {
+    %{$self->SUPER::param_defaults},
+    transcripts_per_batch => 5000,
+    _branch_to_flow_to => 2,
```
[the remark this answers is not in the document]
Comment (ens-carlos) [reply modules/Bio/EnsEMBL/Analysis/Hive/RunnableDB/GenerateGTFLoadingJobs.pm:63]: Applied comment to scallop branch. Could not find this _branch_to_flow_to line.

File: modules/Bio/EnsEMBL/Analysis/Hive/RunnableDB/GenerateGTFLoadingJobs.pm [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    foreach my $chunk (@{$self->process_gtf_file($gtf_file)}){
+      push(@output_ids,{gtf_array => $chunk, logic_name => $logic_name."_rnaseq_gene"});
+    }
+  }
+
+  unless(scalar(@output_ids)) {
+    $self->throw("No output ids created. Something went wrong");
+  }
+  $self->dataflow_output_id(\@output_ids, 2);
+  #$self->output($output_ids);
```
[the remark this answers is not in the document]
Comment (ens-carlos) [reply modules/Bio/EnsEMBL/Analysis/Hive/RunnableDB/GenerateGTFLoadingJobs.pm:126]: Applied comment to scallop branch. Could not find this commented line so it's ok.

File: modules/Bio/EnsEMBL/Analysis/Hive/RunnableDB/HiveCreateStarJobs.pm [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+=cut
+
+package Bio::EnsEMBL::Analysis::Hive::RunnableDB::HiveCreateStarJobs;
+
+use strict;
+use warnings;
+use feature 'say';
+use File::Spec::Functions;
+use Data::Dumper;
```
[the remark this answers is not in the document]
Comment (ens-carlos) [reply modules/Bio/EnsEMBL/Analysis/Hive/RunnableDB/HiveCreateStarJobs.pm:34]: Applied comment to scallop branch. Could not find this line to remove.

File: modules/Bio/EnsEMBL/Analysis/Hive/RunnableDB/HiveCreateStarJobs.pm [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+#  my $output_ids = $self->batch_samples($samples_hash);
+  $self->param('inputlist',$self->batch_samples($samples_hash));
+}
+
+
+sub batch_samples {
+  my ($self,$samples_hash) = @_;
+
+  say Dumper($samples_hash);
```
[the remark this answers is not in the document]
Comment (ens-carlos) [reply modules/Bio/EnsEMBL/Analysis/Hive/RunnableDB/HiveCreateStarJobs.pm:90]: Applied comment to scallop branch. Could not find this line to remove.

File: modules/Bio/EnsEMBL/Analysis/Hive/RunnableDB/HiveCreateStarJobs.pm [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    if(exists $samples_hash->{$sample_id}) {
+      push(@{$samples_hash->{$sample_id}->{'files'}},$result->{$self->param('filename_column')});
+    } else {
+      $samples_hash->{$sample_id}->{'files'} = [$result->{$self->param('filename_column')}];
+      $samples_hash->{$sample_id}->{$self->param('sample_column')} = $sample_name;
+      $samples_hash->{$sample_id}->{$self->param('sample_id_column')} = $sample_id;
+    }
+  }
+
+#  my $output_ids = $self->batch_samples($samples_hash);
```
[the remark this answers is not in the document]
Comment (ens-carlos) [reply modules/Bio/EnsEMBL/Analysis/Hive/RunnableDB/HiveCreateStarJobs.pm:82]: Applied comment to scallop branch. Could not find this line to remove.

File: modules/Bio/EnsEMBL/Analysis/Hive/RunnableDB/HiveStar.pm [hunk 6]
```diff
@@ -1,8 +1,8 @@
 =head1 LICENSE
 
- Copyright [1999-2015] Wellcome Trust Sanger Institute and the EMBL-European Bioinformatics Institute
```
[the remark this answers is not in the document]
Comment (ens-carlos) [reply modules/Bio/EnsEMBL/Analysis/Hive/RunnableDB/HiveStar.pm:3]: Comment applied to scallop branch. Not sure. I think it looks fine in the scallop branch.

File: modules/Bio/EnsEMBL/Analysis/Hive/RunnableDB/HiveStar.pm [hunk 7]
```diff
@@ -46,12 +46,24 @@ package Bio::EnsEMBL::Analysis::Hive::RunnableDB::HiveStar;
...
 
 use parent ('Bio::EnsEMBL::Analysis::Hive::RunnableDB::HiveBaseRunnableDB');
 
 
+sub param_defaults {
+  my ($self) = @_;
+
+  return {
+    %{$self->SUPER::param_defaults},
+    _branch_to_flow_to => 2,
```
[the remark this answers is not in the document]
Comment (ens-carlos) [reply modules/Bio/EnsEMBL/Analysis/Hive/RunnableDB/HiveStar.pm:61]: Applied comment to scallop branch. Could not find this line to remove.

File: modules/Bio/EnsEMBL/Analysis/Hive/RunnableDB/HiveStar.pm [hunk 8]
```diff
@@ -65,26 +77,47 @@ use parent ('Bio::EnsEMBL::Analysis::Hive::RunnableDB::HiveBaseRunnableDB');
...
-  my $filename = $self->param('wide_input_dir').'/'.$self->param('filename');
-  $self->throw("Fastq file $filename not found\n") unless ( -e $filename );
-  my $fastqpair = $self->param('wide_input_dir').'/'.$self->param('fastqpair');
-  $self->throw("Fastq file $fastqpair not found\n") unless ( -e $fastqpair );
-  my $program = $self->param('wide_short_read_aligner');
-  $self->throw("Star program not defined in analysis\n") unless (defined $program);
-  my $runnable = Bio::EnsEMBL::Analysis::Runnable::Star->new
+
+  my $input_ids = $self->param('SM');
+  say "Found ".scalar(@$input_ids)." input ids";
```
[the remark this answers is not in the document]
Comment (ens-carlos) [reply modules/Bio/EnsEMBL/Analysis/Hive/RunnableDB/HiveStar.pm:82]: Applied comment to scallop branch. All says are say_with_header in scallop branch.

File: modules/Bio/EnsEMBL/Analysis/Hive/RunnableDB/HiveStar.pm [hunk 9]
```diff
@@ -65,26 +77,47 @@ use parent ('Bio::EnsEMBL::Analysis::Hive::RunnableDB::HiveBaseRunnableDB');
...
+     -sample_name    => $sample_id,
+     -fastq          => $filepath1,
+     -fastqpair      => $filepath2,
+     -threads        => $self->param('num_threads'),
     );
-  $self->runnable($runnable);
+    $self->runnable($runnable);
+  }
+
+ #  $self->throw("DEBUG");
```
[the remark this answers is not in the document]
Comment (ens-carlos) [reply modules/Bio/EnsEMBL/Analysis/Hive/RunnableDB/HiveStar.pm:120]: Applied comment to scallop branch. Could not find this line to remove.

File: modules/Bio/EnsEMBL/Analysis/Hive/RunnableDB/HiveStar2Introns.pm [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+ Returntype : Hashref
+ Exceptions : None
+
+=cut
+
+sub param_defaults {
+  my ($self) = @_;
+
+  return {
+	  %{$self->SUPER::param_defaults},
```
[the remark this answers is not in the document]
Comment (ens-carlos) [reply modules/Bio/EnsEMBL/Analysis/Hive/RunnableDB/HiveStar2Introns.pm:64]: Applied comment to scallop branch. Alignment is fine in scallop branch.

File: modules/Bio/EnsEMBL/Analysis/Hive/RunnableDB/Stringtie2.pm [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+
+#  my $target_dba = $self->hrdb_get_con("target_db");
+#  my $gene_adaptor = $target_dba->get_GeneAdaptor();
+
+#  my $genes = $self->output();
+#  say "Total genes to output: ".scalar(@$genes);
+#  foreach my $gene (@$genes) {
+#    $gene_adaptor->store($gene);
+#  }
+
```
[the remark this answers is not in the document]
Comment (ens-carlos) [reply modules/Bio/EnsEMBL/Analysis/Hive/RunnableDB/Stringtie2.pm:148]: Applied comment to scallop branch. Removed lines.

File: modules/Bio/EnsEMBL/Analysis/Runnable/Stringtie2.pm [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+  if (defined $val) {
+    throw(ref($val).' is not a Bio::EnsEMBL::DBSQL::DBAdaptor')
+      unless ($val->isa('Bio::EnsEMBL::DBSQL::DBAdaptor'));
+    $self->{_database_adaptor} = $val;
+  }
+
+  return $self->{_database_adaptor};
+}
+
+
```
[the remark this answers is not in the document]
Comment (ens-carlos) [reply modules/Bio/EnsEMBL/Analysis/Runnable/Stringtie2.pm:178]: Comment applied to scallop branch. Already removed.

File: modules/Bio/EnsEMBL/Analysis/Runnable/Stringtie2.pm [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+  my ($self, $val) = @_;
+
+  if ($val) {
+    $self->{_genome_index} = $val;
+  }
+
+  return $self->{_genome_index};
+}
+
+
```
[the remark this answers is not in the document]
Comment (ens-carlos) [reply modules/Bio/EnsEMBL/Analysis/Runnable/Stringtie2.pm:134]: Comment applied to scallop branch. Already removed.

File: modules/Bio/EnsEMBL/Analysis/Runnable/Stringtie2.pm [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+ Description: Run stringtie2 to construct transripts from short reads. The resulting output file will be stored in $self->output
+ Returntype : None
+ Exceptions : None
+
+=cut
+
+sub run {
+  my ($self) = @_;
+
+#  my $genome_index  = $self->genome_index;
```
[the remark this answers is not in the document]
Comment (ens-carlos) [reply modules/Bio/EnsEMBL/Analysis/Runnable/Stringtie2.pm:99]: Comment applied to scallop branch. Removed.

File: modules/Bio/EnsEMBL/Analysis/Runnable/Stringtie2.pm [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+                              Throws if the genome has not been indexed
+
+=cut
+
+sub new {
+  my ( $class, @args ) = @_;
+
+  my $self = $class->SUPER::new(@args);
+  my ($input_file, $num_threads, $output_dir, $delete_input_file) = rearrange([qw (INPUT_FILE NUM_THREADS OUTPUT_DIR DELETE_INPUT_FILE)],@args);
+ # $self->genome_index($genome_index);
```
[the remark this answers is not in the document]
Comment (ens-carlos) [reply modules/Bio/EnsEMBL/Analysis/Runnable/Stringtie2.pm:79]: Comment applied to scallop branch. Removed.

File: modules/Bio/EnsEMBL/Analysis/Hive/RunnableDB/Stringtie2Merge.pm [hunk 16]
```diff
@@ -0,0 +1,225 @@
...
+
+  my $input_file = basename($input_file_path);
+  chomp($input_file);
+
+  my $file_extension_to_remove = $self->param('input_file_extension');
+  if($file_extension_to_remove) {
+    $input_file =~ s/$file_extension_to_remove//;
+  }
+
+  say "Searching for string ".$input_file." in the following csv file:\n".$csv_file;
```
[the remark this answers is not in the document]
Comment (ens-carlos) [reply modules/Bio/EnsEMBL/Analysis/Hive/RunnableDB/Stringtie2Merge.pm:197]: Comment applied to scallop branch. Removed.

File: modules/Bio/EnsEMBL/Analysis/Hive/RunnableDB/Stringtie2Merge.pm [hunk 17]
```diff
@@ -0,0 +1,225 @@
...
+sub fetch_input {
+  my ($self) = @_;
+
+  my $merged_sample_hash = {};
+  $merged_sample_hash->{'merged'} = [];
+
+  my @all_gtf_files = ();
+  my $gtf_dirs = $self->param_required('input_gtf_dirs');
+  foreach my $gtf_dir (@$gtf_dirs) {
+    say "Checking dir: ".$gtf_dir;
```
[the remark this answers is not in the document]
Comment (ens-carlos) [reply modules/Bio/EnsEMBL/Analysis/Hive/RunnableDB/Stringtie2Merge.pm:86]: Comment applied to scallop branch. Removed.

File: modules/Bio/EnsEMBL/Analysis/Hive/RunnableDB/Stringtie2Merge.pm [hunk 18]
```diff
@@ -0,0 +1,225 @@
...
+  foreach my $tissue (@tissues) {
+    my @tissue_files = @{$merged_sample_hash->{$tissue}};
+    unless(scalar(@tissue_files)) {
+      $self->throw("There were no files for the tissue. This should not happen. Tissue affected: ".$tissue);
+    }
+
+    my $merged_file_path = catfile($stringtie_merge_dir,$tissue.".gtf");
+
+    if(scalar(@tissue_files) == 1) {
+      say "For ".$tissue." only found one file so will just copy it to merged dir";
```
[the remark this answers is not in the document]
Comment (ens-carlos) [reply modules/Bio/EnsEMBL/Analysis/Hive/RunnableDB/Stringtie2Merge.pm:153]: Comment applied to scallop branch. Replaced.

File: modules/Bio/EnsEMBL/Analysis/Hive/RunnableDB/Stringtie2.pm [hunk 19]
```diff
@@ -0,0 +1,192 @@
...
+  my $input_file = basename($input_file_path);
+  chomp($input_file);
+
+  my $file_extensions_to_remove = $self->param('input_file_extensions');
+
+  foreach my $extension (@$file_extensions_to_remove) {
+    $input_file =~ s/$extension//;
+  }
+
+  say "Searching for string ".$input_file." in the following csv file:\n".$csv_file;
```
[the remark this answers is not in the document]
Comment (ens-carlos) [reply modules/Bio/EnsEMBL/Analysis/Hive/RunnableDB/Stringtie2.pm:164]: Comment applied to scallop branch. Replaced.

File: modules/Bio/EnsEMBL/Analysis/Hive/RunnableDB/Stringtie2.pm [hunk 20]
```diff
@@ -0,0 +1,192 @@
...
+
+    # Here we should have found the line, so the sample name should be in the first column
+    unless($line =~ /^([^\t]+)\t/) {
+      $self->throw("Tried to parse the sample name from corresponding line in the csv file but failed (parsing regex issue?). Line:\n".$line);
+    }
+
+    $sample_name = $1;
+    last;
+  }
+  close IN;
```
[the remark this answers is not in the document]
Comment (ens-carlos) [reply modules/Bio/EnsEMBL/Analysis/Hive/RunnableDB/Stringtie2.pm:187]: Comment applied to scallop branch. Throw if file cannot be closed.

File: modules/Bio/EnsEMBL/Analysis/Hive/RunnableDB/Stringtie2Merge.pm [hunk 21]
```diff
@@ -0,0 +1,225 @@
...
+
+    my $tissue_file_list_path = $merged_file_path.".list";
+    unless(open(OUT,">".$tissue_file_list_path)) {
+      $self->throw("Failed to open output file to list gtf files for merge in. Path used:\n".$tissue_file_list_path);
+    }
+
+    foreach my $file_path (@tissue_files) {
+      say OUT $file_path;
+    }
+    close OUT;
```
[the remark this answers is not in the document]
Comment (ens-carlos) [reply modules/Bio/EnsEMBL/Analysis/Hive/RunnableDB/Stringtie2Merge.pm:170]: Comment applied to scallop branch. Throw if file cannot be closed.

File: modules/Bio/EnsEMBL/Analysis/Hive/RunnableDB/Stringtie2Merge.pm [hunk 22]
```diff
@@ -0,0 +1,225 @@
...
+
+    # Here we should have found the line, so the sample name should be in the first column
+    unless($line =~ /^([^\t]+)\t/) {
+      $self->throw("Tried to parse the sample name from corresponding line in the csv file but failed (parsing regex issue?). Line:\n".$line);
+    }
+
+    $sample_name = $1;
+    last;
+  }
+  close IN;
```
[the remark this answers is not in the document]
Comment (ens-carlos) [reply modules/Bio/EnsEMBL/Analysis/Hive/RunnableDB/Stringtie2Merge.pm:220]: Comment applied to scallop branch. Throw if file cannot be closed.

**Comment by thibauthourlier** [comment]
I'm closing the PR as it is better to resubmit it when the Star/Scallop config has been verified

**Pull request closed without merging**